### DISCUSSION TITLE
MTB 23.5, error SQL when start program

### DISCUSSION THREAD

**Discussion by nadine-90** [mytourbook/mytourbook#1145] [General]
I have always used MTB version 18, now I use windows 11 and have installed MTB version 23.5. But when I start the application, I get this error (see image).
I would like to continue using this nice program, could you help me? Thank you.
![MyTB](https://github.com/mytourbook/mytourbook/assets/138731437/af2ee41b-57c6-44d4-bfca-22df6b0aeaf8)

**Comment by wolfgang-ch** [comment]
Can you attach here these log files

C:\Users\<YOUR-NAME>\mytourbook\.metadata\.log
C:\Users\<YOUR-NAME>\mytourbook\.metadata\.bak_0.log

C:\Users\<YOUR-NAME>\mytourbook\derby-database\derby.log

**Comment by nadine-90** [comment]
This morning the program started without problems, or anyway without errors. However, I previously deleted the Mytourbook folder in my user, which was the one copied to the operating system where I'm using MTB 18.7.
I attach the files anyway (bak file doesn't exist for now).
But now I would like to know, how can I import at least my routes done in recent years into the MTB 23.5 database? For settings-preferences it doesn't matter.
Thank you
[derby.log](https://github.com/mytourbook/mytourbook/files/11966379/derby.log)
[metadata_log.log](https://github.com/mytourbook/mytourbook/files/11966380/metadata_log.log)

**Comment by wolfgang-ch** [comment]
> But now I would like to know, how can I import at least my routes done in recent years into the MTB 23.5 database? 

You can replace the tour database by replacing the whole content of this folder `C:\Users\<YOUR-NAME>\mytourbook\derby-database`. 

I do this very often **but in the reverse order** when I copy my production tour db into my dev environment with this script
```
rmdir workspace\derby-database /q/s
xcopy "C:\Users\Wolfgang\mytourbook\derby-database" workspace\derby-database\ /S
```

**Comment by nadine-90** [comment]
If I copy the entire contents of the MTB 18.7 derby-database folder and replace the contents of the MTB 23.5 derby-database folder, I get the same error as in the first message above. Thank you

**Comment by wolfgang-ch** [comment]
Can you attach the log files from the location from where you copied the tour database. I need to exactly know which exact database version you were using because the above log files are from the new installation and do not show this info.

**Comment by nadine-90** [comment]
I deleted the data log just before this test; I then logged in without importing anything to MTB 18.7, and it went fine. I then copied the contents of the MTB 18.7 'derby database' folder into the MTB 23.5 derby database folder and MTB crashes as soon as I start it.

[.log](https://github.com/mytourbook/mytourbook/files/11983062/default.log)
[derby.log](https://github.com/mytourbook/mytourbook/files/11983064/derby.log)
![start MTB error](https://github.com/mytourbook/mytourbook/assets/138731437/a81a43ec-12b3-4981-b2b4-1f6c95940c95)

I also did 'refresh database', the program starts, but the imported paths are not there

Thanks

**Reply by wolfgang-ch** [reply]
The second screenshot shows the info which I wanted. Tomorrow I will test if I can reproduce this issue.

**Comment by nadine-90** [comment]
Thanks for your help :-)

**Comment by wolfgang-ch** [comment]
I could reproduce this error by 

* Install MT 18.7
* Run MT
* Import/save some tours (is propably not needed)
* Install MT 23.5
* Run MT -> the error occurs

I'll need to fix this before you can use MT 23.5 with your saved tours, however the old MT 18.7 should still work.


```
09:57:48.994[Splash message] Starting Application...
Application started in : 1467ms
09:57:49.964[Splash message] Starting Database...
09:57:48.848[Executing Derby command] jdbc:derby:tourbook;create=true
09:57:48.848[Executing Derby command] jdbc:derby:tourbook
09:57:48.848[Executing Derby command] jdbc:derby:tourbook
09:57:48.848[DB Update] Current database DESIGN version: 35
09:57:48.848[DB Update] DB upgrade BEFORE db design update is needed 35
09:57:48.848[Executing Derby command] jdbc:derby:;shutdown=true
09:57:48.848[Derby] Derby server is shutdown
09:57:48.848[Executing Derby command] jdbc:derby:tourbook;upgrade=true
09:57:49.407[Splash message] Upgrading database... 

!ENTRY net.tourbook.common 4 0 2023-07-08 09:57:49.508
!MESSAGE SQLException

SQLState: 08001
ErrorCode: 0
Message: No suitable driver found for jdbc:derby:tourbook;upgrade=true
java.sql.SQLException: No suitable driver found for jdbc:derby:tourbook;upgrade=true
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:706)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:229)
	at net.tourbook.database.TourDatabase.sqlStartup_UpgradeDerbyDatabase(TourDatabase.java:5552)
	at net.tourbook.database.TourDatabase.sqlStartup_UpgradedDb_1_BeforeDbDesignUpdate(TourDatabase.java:5482)
	at net.tourbook.database.TourDatabase.sqlStartup_50_IsDesignVersionValid(TourDatabase.java:5336)
	at net.tourbook.database.TourDatabase.lambda$6(TourDatabase.java:5005)
	at net.tourbook.database.TourDatabase.sqlStartup_10_IsSqlServerUpAndRunning(TourDatabase.java:5037)
	at net.tourbook.database.TourDatabase.getEntityManager(TourDatabase.java:4824)
	at net.tourbook.database.PersonManager.getPeopleFromDb(PersonManager.java:61)
	at net.tourbook.database.PersonManager.getTourPeople(PersonManager.java:113)

```

**Comment by wolfgang-ch** [comment]
This bug occurs when the old version is MT 18.7 and the `Database System` is `Embedded Server`. When the `Database System` is `Network server` then the update for MT 23.5 works.

![image](https://github.com/mytourbook/mytourbook/assets/1283445/ab72af2d-40d8-48f3-96b0-4d844b43c07b)

Here are two scenarios which can fix your issue

**For both scenarios ensure, that these folders are deleted before installing MT:** 

C:\Program Files\MyTourbook
C:\Program Files (x86)\MyTourbook

# Scenario I

* Install Java 8             https://adoptium.net/temurin/archive/?version=8
* Install + run MT 18.7      https://sourceforge.net/projects/mytourbook/files/MyTourbook/18.7.0/
* Select Network server + restart MT

![image](https://github.com/mytourbook/mytourbook/assets/1283445/f63c78a3-d7bb-4fba-93ec-94abd6143bfc)

* Install Java >=11          https://adoptium.net/temurin/archive/?version=17
* Install + run MT 23.5      https://sourceforge.net/projects/mytourbook/files/MyTourbook/23.5.0/

![image](https://github.com/mytourbook/mytourbook/assets/1283445/41e927fe-eb7a-4ed2-9b44-ce21272fc1a7)

# Scenario II 

* Install Java 8             https://adoptium.net/temurin/archive/?version=8
* Install + run MT 18.7      https://sourceforge.net/projects/mytourbook/files/MyTourbook/18.7.0/
* Install + run MT 18.13     https://sourceforge.net/projects/mytourbook/files/MyTourbook/18.13.0/
* Install Java >=11          https://adoptium.net/temurin/archive/?version=17
* Install + run MT 23.5      https://sourceforge.net/projects/mytourbook/files/MyTourbook/23.5.0/


# Errors

1. Ignore this error message when it is displayed

![image](https://github.com/mytourbook/mytourbook/assets/1283445/5639f503-f524-4eaa-ad49-43b7c500f42a)

2. Install MT 23.5.1 when the import view shows an error

https://github.com/mytourbook/mytourbook/issues/1133
https://mytourbook.github.io/update-sites/23.5.1/

**Comment by nadine-90** [comment]
Scenario 1 seems easier to me, I'll try that. Thanks, I'll update the situation here in the forum

**Comment by nadine-90** [comment]
I did everything: MTB 18.7, java 8, MTB 23.5, java 11.. and ok open. But the link to install MTB 23.5.1 doesn't exist... (yes, I have the error in importing the routes..).
Thanks

**Comment by wolfgang-ch** [comment]
>I did everything: MTB 18.7, java 8, MTB 23.5, java 11.. and ok open.

It's great to here that this worked 👍  

> But the link to install MTB 23.5.1 doesn't exist... 

This link is for the [23.5.1 update site.](https://mytourbook.github.io/update-sites/23.5.1/) 

How to install update sites is documented in https://mytourbook.sourceforge.io/mytourbook/index.php/download-install/software-updates

**Comment by nadine-90** [comment]
Update 23.5.1 fails
![update](https://github.com/mytourbook/mytourbook/assets/138731437/ed22e511-27a8-4d77-b429-14f855e0dec2)

Thank you Wolfgang

[.log](https://github.com/mytourbook/mytourbook/files/11995323/default.log)

**Reply by wolfgang-ch** [reply]
Have you run MT as administrator when doing the update?

![image](https://github.com/mytourbook/mytourbook/assets/1283445/6458268d-e78d-4479-9edf-daf7edf3f002)

**Comment by nadine-90** [comment]
Ok wolf, doing as administration I can go further; but then I get to the error you see in the image. As per the guide it says to repeat the procedure, but I've done it several times, but always the same error. Do we have any solution? Thank you

![error](https://github.com/mytourbook/mytourbook/assets/138731437/cf1be9e0-ba54-4b1d-8660-ce7efa5a3a04)

**Comment by wolfgang-ch** [comment]
You can install 23.5.1 also with this procedure:

* Download https://github.com/mytourbook/mytourbook.github.io/blob/main/update-sites/23.5.1.zip
* Expand this zip file in any location
* Do the update process as described in the documentation https://mytourbook.sourceforge.io/mytourbook/index.php/download-install/software-updates until this dialog is opened
![image](https://github.com/mytourbook/mytourbook/assets/1283445/addc6766-07d0-41d9-aea8-8d2c6bec72fc)
* Select the `Add...` button
* Add the expanded 23.5.1 folder with these 2 buttons
![image](https://github.com/mytourbook/mytourbook/assets/1283445/80029783-154c-46dd-975a-8b0b419839d2)
* The further steps are the same as in the documentation

**Comment by nadine-90** [comment]
Thanks wolf, now the update is installed and the program starts normally. All I have left is importing the tours I have in the old operating system; I tried to change the contents of the tourbook folder only, but it warns me that I should update the database from 35 to 50 (but it's already been done, right?). Maybe I need to copy/paste different content? You told me the entire derby-database folder, but it seems to me that in this version there are new files different from the old version. Take one more small step and I will live happily ever after. Thank you

**Comment by wolfgang-ch** [comment]
> Thanks wolf, now the update is installed and the program starts normally. All I have left is importing the tours I have in the old operating system; I tried to change the contents of the tourbook folder only, but it warns me that I should update the database from 35 to 50 (but it's already been done, right?). 

The version number 35 or 50 is saved in the database, so the database you are using is not yet updated.

>Maybe I need to copy/paste different content? You told me the entire derby-database folder, but it seems to me that in this version there are new files different from the old version. 

Don't touch any files in the folder `...\derby-database\tourbook`, the only thing you can do is delete `...\derby-database\lucene-index` when the fulltext index is corrupt

![image](https://github.com/mytourbook/mytourbook/assets/1283445/ec5d0ab0-4e54-424d-b2d5-491c75c41d9d)

![image](https://github.com/mytourbook/mytourbook/assets/1283445/9d1d37fd-0401-4f3f-8d02-887b300493a5)


>Take one more small step and I will live happily ever after. Thank you

You should replace the derby-database folder with the folder which you have updated. I don't know your setup and where you have saved/backup your files.

**Comment by nadine-90** [comment]
I tried all possibilities to import routes from the old version, but I was unsuccessful. I made the new settings and imported 300 routes again: fast and problem-free operation. It is not necessary for me to import anything from the old version. Thanks for the support.
